_posts/2019-10-01-intro-to-hmc/hmc-intro.Rmd

```diff
@@ -0,0 +1,284 @@
+  current_K <- sum(current_p^2) / 2
+  proposed_U <- U(q)
+  proposed_K <- sum(p^2) / 2
+  # Accept or reject the state at end of trajectory, returning either
+  # the position at the end of the trajectory or the initial position
+  if (runif(1) < exp(current_U-proposed_U+current_K-proposed_K)) {
+    return (q)  # accept
+  } else {
+    return (current_q)  # reject
+  }
+}
+```
+
+Hopefully, you find this piece of code as helpful as I do. Are we through yet? Well, so far we haven't encountered the last magic word: NUTS. What, or who, is NUTS?
+
+## NUTS
+
+NUTS, [added to Stan in 2011](https://statmodeling.stat.columbia.edu/2011/11/30/stan-uses-nuts/) ^[As of today, Stan uses a modified version, see appendix A.5 of [@2017arXiv170102434B], to be cited soon.] and about a month ago, to [TensorFlow Probability's master branch](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/mcmc/nuts.py), is an algorithm that aims to circumvent one of the practical difficulties in using HMC: The choice of number of leapfrog steps to perform before making a proposal. The acronym stands for No-U-Turn Sampler, alluding to the avoidance of U-turn-shaped curves in the optimization landscape when the number of leapfrog steps is chosen too high. 
+
+The reference paper by Hoffman & Gelman [@hoffman2011nouturn] also describes a solution to a related difficulty: choosing the step size $\epsilon$. The respective algorithm, _dual averaging_, [was also recently added to TFP](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/mcmc/dual_averaging_step_size_adaptation.py).
+
+NUTS being more of algorithm in the computer science usage of the word than a thing to explain conceptually, we'll leave it at that, and ask the interested reader to read the paper -- or even, [consult the TFP documentation to see how NUTS is implemented there](https://github.com/tensorflow/probability/blob/master/discussion/technical_note_on_unrolled_nuts.md). Instead, we'll round up with another conceptual analogy, Michael Bétancourts crashing (or not!) satellite [@2017arXiv170102434B].
+
+## How to avoid crashes
+
+Bétancourt's article is an awesome read, and a paragraph focusing on a single point made in the paper can be nothing than a "teaser" (which is why we'll have a picture, too!). 
+
+To introduce the upcoming analogy, the problem starts with high dimensionality, which is a given in most real-world problems. In high dimensions, as usual, the density function has a _mode_ (the place where it is maximal), but necessarily, there cannot be much _volume_ around it -- just like with k-nearest neighbors, the more dimensions you add, the farther your nearest neighbor will be.
+A product of volume and density, the only significant probability mass resides in the so-called _typical set_ ^[for a detailed discussion of typical sets, see the book by McKay [@MacKay:2002:ITI:971143].] , which becomes more and more narrow in high dimensions.
```

Comment:
Might be worth adding a sentence or two on what typical sets are (if that's epistemically possible), since we're using the term down the post and people aren't likely going to acquire the book to read up on it

Reply:
yeah I see... I added a link to Wikipedia :-)